File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <bits/stdc++.h>

using namespace std;

bool comparePlays(pair<int,int> &a, pair<int,int> &b){ //고유번호, 재생수
    if(a.second == b.second) return a.first < b.first; //고유번호 오름차순
    else return a.second > b.second; //재생 수 내림차순
}
bool compareGenres(pair<string,int> &a, pair<string,int> &b){//장르, 총 재생수
    return a.second > b.second; //총 재생 수 내림차순
}

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    unordered_map<string, int> totalPlays; //장르별 총 재생수
    unordered_map<string, vector<pair<int,int>>> genrePlays; //곡별 재생 수
    
    for(int i=0;i<genres.size();i++){
        totalPlays[genres[i]] += plays[i];
        genrePlays[genres[i]].push_back(make_pair(i,plays[i]));
    }
    
    vector<pair<string, int>> total(totalPlays.begin(), totalPlays.end());
    sort(total.begin(), total.end(), compareGenres);
    
    for(const auto& t : total){
        string genre = t.first;
        sort(genrePlays[genre].begin(), genrePlays[genre].end(),comparePlays);
    }
    
    for(const auto& t : total){
        string genre = t.first;
        if(genrePlays[genre].size() > 1) for(int i=0; i<2; i++) answer.push_back(genrePlays[genre][i].first);
        else answer.push_back(genrePlays[genre][0].first);
    }
    
    return answer;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    
    genre_num = dict()
    genre_plays = dict()
    
    for i in range(len(genres)):
        genre_num[genres[i]] = genre_num.get(genres[i], []) + [i]
        genre_plays[genres[i]] = genre_plays.get(genres[i], 0) + plays[i]
        
    # 많이 재생된 장르 순
    genre_sort = sorted(list(set(genres)), key = lambda x: -genre_plays[x])
    print(genre_sort)
    answer = []
    for genre in genre_sort:
        if len(genre_num[genre]) == 1:
            answer += genre_num[genre]
        else:    
            song_sort = sorted(genre_num[genre], key=lambda x: [plays[x], -x], reverse=True)
            answer += song_sort[:2]
        
    return answer 